Extract site flag queries into Apollo and set up preloader for Layout

philipp-spiess opened the pull request:
This PR updates the site flag observable based propagation to Apollo so we can properly preload the data and avoid having to subscribe to a global observable.

## Test plan

- You can test the refetching behavior by e.g. changing a site admin config page in the admin panel

https://user-images.githubusercontent.com/458591/222157750-d5b03217-456f-4d2f-baac-7e35ed5be260.mov

- I also ensured that the emails settings page still worked as expected.


<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

## App preview:

- [Web](https://sg-web-ps-clean-up-site-flags.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

philipp-spiess commented on client/web/src/global/GlobalAlerts.tsx:
```diff
@@ -22,20 +22,49 @@ import { Notices, VerifyEmailNotices } from './Notices'
-    const siteFlagsValue = useObservable(siteFlags)
+export const GlobalAlerts: React.FunctionComponent<Props> = ({ authenticatedUser, isSourcegraphDotCom }) => {
+    const settings = useSettings()
+    const { data } = useQuery<GlobalAlertsSiteFlagsResult, GlobalAlertsSiteFlagsVariables>(QUERY, {})
```
@valerybugakov What's a good way to make this preloaded but still allow this component to be mounted in the `LegacyLayout` and be lazy loaded there? 😬 I think we forgot to think that through?

I want to avoid forking it just so that we have one preloaded version 🤔  We could hoist only the query so we have one version in the Layout.loader.ts but then we need a copy in `legacylayout` which also doesn't work (we can't have the same query twice) so we'd need to import the `Layout.loader` from the `LegacyLayout`.

valerybugakov replied:
Thanks to the work you've done here on migrating these queries to Apollo Client, it should be relatively straightforward. This component can remain the same in the legacy version with one change: `fetchPolicy: 'cache-first'`. If we preload site flags in `Layout.loader.ts`, then `GlobalAlerts` in the Storm version won't issue an API request on the page load because the cache will already be filled with the required fields. What do you think?

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.03% (-0.84 kb) | 0.00% (+0.22 kb) | 0.01% (+1.06 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ps/clean-up-site-flags/compare-report.html;tab=live_object) for a full comparison between the commits 13e7011 and b17e8c3 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

valerybugakov commented on client/web/src/site/backend.ts:
```diff
@@ -1,71 +1,10 @@
- * refreshSiteFlags refreshes the site flags. The result is available from
- * the siteFlags const.
- */
-export function refreshSiteFlags(): Observable<never> {
-    return authRequired.pipe(
-        take(1),
-        filter(authRequired => !authRequired),
-        mergeMap(() =>
-            requestGraphQL<SiteFlagsResult, SiteFlagsVariables>(gql`
```
Yes, please!

philipp-spiess commented:
@valerybugakov I went ahead and added the preloader stuff in this PR as well. Seems like it works, the site query is being cached now :)

I think we can do the same with `DefaultSearchContextSpec` so in #48472 I'll also migrate that to use Apollo. 

One question though, maybe I’m doing something wrong but the feature flag preloading that you added in SearchPage doesn't seem to be working like I expect it to? I added preloading for the three flags used in the navbar yet the graphql queries are still issued:

<img width="1318" alt="Screenshot 2023-03-02 at 18 51 15" src="https://user-images.githubusercontent.com/458591/222512208-206457d1-5448-4780-8e46-d406d04b9f0c.png">

<img width="1702" alt="Screenshot 2023-03-02 at 18 54 25" src="https://user-images.githubusercontent.com/458591/222512239-7c305021-37cd-4a0c-9fc2-fe4a3bda7433.png">

Is there still something missing or am I misunderstanding?

philipp-spiess commented:
@valerybugakov Do you mind taking another quick look at the changes? This should be good to merge now!

valerybugakov commented on client/web/src/user/settings/emails/UserSettingsEmailsPage.tsx:
```diff
@@ -26,11 +32,25 @@ type UserEmail = (NonNullable<UserEmailsResult['node']> & { __typename: 'User' }
     const [statusOrError, setStatusOrError] = useState<Status>()
     const [emailActionError, setEmailActionError] = useState<EmailActionError>()
 
+    const { data } = useQuery<UserSettingsEmailsSiteFlagsResult, UserSettingsEmailsSiteFlagsVariables>(FLAGS_QUERY, {})
```
I do not understand why did they change the `OperationVariables` type to `Record<string, any>` and now these empty objects are required 😬

valerybugakov reviewed (approved):
Yay! Fully preloaded layout 🎉

I noticed just one issue testing these changes locally: when I sign out, 401 status errors are propagated to the `ErrorBoundary`, and instead of seeing the login form, I see the error message screen. It doesn't happen if the storm is not enabled. Do you have ideas on where this is coming from?

philipp-spiess commented:
@valerybugakov Ugh I don't even know how this worked before 🙃
